File: leetcode/exercises/2.1.4.medium.container_with_most_water.go
// -----------------------
// Created     :3/11/2025
// Last Edited :3/11/2025 
// Topics      :
// Big O       :
// Problem Id  :1
// Source      :
// Notes       :
// -----------------------

package main

import (
    "fmt"
)

func maxArea(height []int) int {
    l := 0
    r := len(height) - 1
    max_height := 0
    base := 0
    h := 0
    for i := 0; i < len(height); i++ {
        base = max((r - l), -(r - l))
        h = min(height[l], height[r])
        max_height = max(max_height, base*h)
        if height[l] < height[r] {
            l++
        } else if height[l] > height[r] {
            r--
        } else {
            l++
            r--
        }
    }
    return max_height
}

func main() {
    fmt.Print(maxArea([]int{1,2,1}));
}

File: leetcode/exercises/2.1.4.medium.container_with_most_water.py
# -----------------------
# Created     : 13/11/2025
# Last Edited : 13/11/2025 
# Topics      : 
# Big O       :
# Problem Id  : 11
# References  :
# Notes       :
# -----------------------

class Solution:
    def maxArea(self, height: List[int]) -> int:
        if len(height) < 2: return 0

        l = 0
        r = len(height) - 1
        max_area = 0
        while l < r:
            base = abs(r - l)
            min_height = min(height[l], height[r])
            max_area = max(max_area, base*min_height)

            if height[l] < height[r]:
                 l += 1
            elif height[l] > height[r]:
                 r -= 1
            else:
                 r -= 1
                 l += 1
        return max_area

s = Solution()
print(s.maxArea([1,2,1]))

            

        
